sympy/physics/quantum/state.py

```diff
@@ -617,6 +619,60 @@ def dual_class(self):
         return TimeDepKet
 
 
+class OrthogonalState(State, StateBase):
+    """General abstract quantum state used as a base class for Ket and Bra."""
+    pass
+
+class OrthogonalKet(OrthogonalState, KetBase):
+    """Orthogonal Ket in quantum mechanics.
+
+    The inner product of two states with different labels will give zero,
+    states with the same label will give one.
+
+        >>> from sympy.physics.quantum import OrthogonalBra, OrthogonalKet, qapply
+        >>> from sympy.abc import m, n
+        >>> (OrthogonalBra(n)*OrthogonalKet(n)).doit()
+        1
+        >>> (OrthogonalBra(n)*OrthogonalKet(n+1)).doit()
+        0
+        >>> (OrthogonalBra(n)*OrthogonalKet(m)).doit()
+        <n|m>
+    """
+
+    @classmethod
+    def dual_class(self):
+        return OrthogonalBra
+
+    def _eval_innerproduct(self, bra, **hints):
+
+        if len(self.args) != len(bra.args):
+            raise ValueError('Cannot multiply a ket that has a different number of labels.')
+
+        for i in range(len(self.args)):
+            diff = self.args[i] - bra.args[i]
+            diff = diff.simplify()
+
+            if diff.is_nonzero:
+                return 0
+
+            if not diff.is_zero:
```

Comment:
I don't understand what this is doing. The `is_nonzero` property is only true for real values:
```
>>> sqrt(-1).is_nonzero
False
```
Both `is_zero` and `is_nonzero` use fuzzy logic so can be True, False or None. That means that `not diff.is_zero` will give True if `diff.is_zero` is False or None.

Reply:
Then I obviously also didn't understand what I were doing. But somehow it works... ?
I'll rewrite the code only in terms of `is_zero`.